### PR TITLE
Annotate EqQuery's methods nullable and apply missing null checks

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
EqQuery's methods are required to return `nulls` to signal the query needs to fall back to GenericFunctionQuery (or to the unoptimized query).
This PR annotates the methods and applies missing null checks.

This is a bug in `fix-lucene-query-builder` branch as it has never been allowed to query unindexed, no-doc-values columns, a change entry is not required.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java [hunk 1]
```diff
@@ -682,4 +682,13 @@ public void test_eq_object_with_undefined_key() {
         Query query = convert("obj = {x=1, y=2, z=3}"); // z undefined
         assertThat(query).hasToString("+obj.x:[1 TO 1] +obj.y:[2 TO 2] #(obj = {\"x\"=1, \"y\"=2, \"z\"=3})");
     }
+
+    @Test
+    public void test_query_on_double_array_with_index_off_and_no_docvalues_falls_back_to_generic_query() {
+        Query query = convert("d_array_index_off_no_docvalues[1] = 12.34");
+        assertThat(query).isExactlyInstanceOf(GenericFunctionQuery.class);
```
Comment (jeeminso) [inline server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java:689]: This covers for the changes to `SubscriptFunction`.

File: server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java [hunk 2]
```diff
@@ -682,4 +682,13 @@ public void test_eq_object_with_undefined_key() {
...
+
+        query = convert("12.34 != any(d_array_index_off_no_docvalues)");
+        assertThat(query).isExactlyInstanceOf(GenericFunctionQuery.class);
```
Comment (jeeminso) [inline server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java:692]: This covers the changes to `AnyNeqOperator`.

File: server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java [hunk 3]
```diff
@@ -108,27 +108,26 @@ protected Query literalMatchesAnyArrayRef(Function any, Literal<?> probe, Refere
...
+            candidates.indexType() != IndexType.NONE);
+        var lt = eqQuery.rangeQuery(
+            columnName,
+            null,
+            value,
+            false,
+            false,
+            candidates.hasDocValues(),
+            candidates.indexType() != IndexType.NONE);
+        if (lt == null || gt == null) {
```
Comment (matriv) [inline server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java:126]: I guess that both should be null if we cannot create a rangeQuery for lt, than we cannot create also for gt, right? So maybe use `&&` and add an `assert` to verify that if one is null both are null?

File: server/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 
 package io.crate.execution.engine.collect.collectors;
 
+import static org.elasticsearch.cluster.metadata.Metadata.COLUMN_OID_UNASSIGNED;
```
Comment (matriv) [inline server/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java:24]: Unused imports? (also below)

File: server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java [hunk 5]
```diff
@@ -682,4 +682,13 @@ public void test_eq_object_with_undefined_key() {
         Query query = convert("obj = {x=1, y=2, z=3}"); // z undefined
         assertThat(query).hasToString("+obj.x:[1 TO 1] +obj.y:[2 TO 2] #(obj = {\"x\"=1, \"y\"=2, \"z\"=3})");
     }
+
+    @Test
+    public void test_query_on_double_array_with_index_off_and_no_docvalues_falls_back_to_generic_query() {
```
Comment (matriv) [inline server/src/test/java/io/crate/lucene/CommonQueryBuilderTest.java:687]:
```suggestion
    public void test_equality_query_on_double_array_with_index_off_and_no_docvalues_falls_back_to_generic_query() {
```

File: server/src/test/java/io/crate/execution/engine/collect/collectors/OptimizeQueryForSearchAfterTest.java [hunk 6]
```diff
@@ -88,4 +88,25 @@ public void test_short_range_query_with_and_without_docvalues() {
         assertThat(booleanQuery.clauses()).satisfiesExactly(
                 x -> assertThat(x.getQuery()).isNotInstanceOf(IndexOrDocValuesQuery.class));
     }
+
+    @Test
+    public void test_short_range_query_without_doc_values_and_unindexed_return_null() {
```
Comment (matriv) [inline server/src/test/java/io/crate/execution/engine/collect/collectors/OptimizeQueryForSearchAfterTest.java:93]:
```suggestion
    public void test_range_query_on_short_without_doc_values_and_unindexed_return_null() {
```

**Review by matriv — APPROVED** [review]
Thx, left some minor comments

File: server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java (hunk 3 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java:126]: Thank you! updated

File: server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java [hunk 7]
```diff
@@ -108,27 +108,28 @@ protected Query literalMatchesAnyArrayRef(Function any, Literal<?> probe, Refere
...
+            candidates.indexType() != IndexType.NONE);
+        var lt = eqQuery.rangeQuery(
+            columnName,
+            null,
+            value,
+            false,
+            false,
+            candidates.hasDocValues(),
+            candidates.indexType() != IndexType.NONE);
+        assert lt != null || gt == null : "If lt is null, gt must be null";
```
Comment (matriv) [inline server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java:127]: This assertion should inside the if below with the `&&`, before `return null` like this:
```
assert lt != null || gt != null : "Both gt and lt should be null";
```

Comment (jeeminso) [reply server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java:127]: Wouldn't it be already `lt == null` such that `assert lt != null ...` be redundant?

Comment (matriv) [reply server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java:127]: yep, sorry, the `if` should also become an `||`, so you get in if one of the 2 is is null, then you assert that both should be null, and then `return null`.

**Pull request merged**